src/main/java/org/jenkinsci/plugins/envinject/service/EnvInjectEnvVars.java

```diff
@@ -218,9 +217,16 @@ public void resolveVars(Map<String, String> variables, Map<String, String> env)
             int previousNbUnresolvedVar = nbUnresolvedVar;
             nbUnresolvedVar = 0;
             for (Map.Entry<String, String> entry : variables.entrySet()) {
-                String value = Util.replaceMacro(entry.getValue(), variables);
-                entry.setValue(value);
-                if (isUnresolvedVar(value)) {
+                String key = entry.getKey();
+                String value = entry.getValue();
+                Map<String, String> resolveVariables = new HashMap<String, String>(variables);
+                // avoid self reference expansion
+                // FOO=$FOO-X -> FOO=$FOO-X-X -> FOO=$FOO-X-X-X etc.
+                resolveVariables.remove(key);
+                String newValue = Util.replaceMacro(value, resolveVariables);
+                entry.setValue(newValue);
+                boolean variableChanged = !newValue.equals(value);
+                if (variableChanged && isUnresolvedVar(newValue)) {
```

Comment:
i think it is not required, but i felt it is more correct.
In the loop we resolve variables against themselves... if a variable's value did not change by resolving and it is still not resolved yet, it wont resolve in any of the future iterations, thus we should not consider it unresolved with regard to the continuation of the loop.

In practice it probably just saves a single loop at the end.
So if you prefer to keep the previous logic, i can change it.